# Profitable App Profiles for the App Store and Google Play Markets



**Introduction**

This project will involve working as a data analyst for a company that builds Android and iOS mobile apps. The apps are available on Google Play and the App Store. The goal of this project is to analyze data to help our developers understand what kinds of apps are likely to attract more users.




In [1]:
##Opening both App store data and Google Play Data

from csv import reader

opened_file = open('AppleStore.csv')
read_file = reader(opened_file)
app_data = list(read_file)

opened_file = open('googleplaystore.csv')
read_file = reader(opened_file)
gps_data =list(read_file)



## Exploring the Data Set


In [2]:

def explore_data(dataset, start, end, row_and_columns=False):
    dataset_slice = dataset[start:end]
    for row in dataset_slice:
        print(row)
        print('\n')
        
    if row_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))
        
#Printing 4 rows, without headers.
print('Apple Store Data \n')
app_print = explore_data(app_data, 1, 5, False)
print('\n Google Play Store Data \n')
gps_print = explore_data(gps_data, 1, 5, False)


#Finding number of rows and columns
print('Apple Store Data - Columns / Length \n')
app_i = explore_data(app_data, 0, 1, True)
print('\n Google Play Store Data - Columns / Length \n')
gps_i = explore_data(gps_data, 0, 1, True)



Apple Store Data 

['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


['420009108', 'Temple Run', '65921024', 'USD', '0.0', '1724546', '3842', '4.5', '4.0', '1.6.2', '9+', 'Games', '40', '5', '1', '1']



 Google Play Store Data 

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Li

## Deleting Incorrect Data


In [3]:
#Finding the error on 10472
print(gps_data[0])
print('\n')
print(gps_data[10473])
print('\n')

print(gps_data[2])


['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


Row 10473 has a issue where the rating is 19, we can clearly delete this row as the maximum for the rating can only be 5

In [4]:
del gps_data[10473]

print(gps_data[10473])

['osmino Wi-Fi: free WiFi', 'TOOLS', '4.2', '134203', '4.1M', '10,000,000+', 'Free', '0', 'Everyone', 'Tools', 'August 7, 2018', '6.06.14', '4.4 and up']


## Removing Duplicate Entries

In [5]:
##printing duplicates on instagram in the google playstore data


for app in gps_data[1:]:
    name = app[0]
    if name == 'Instagram':
        print(app)

['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577446', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66509917', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


In [7]:
## Finding number of occurences of duplicate names


duplicate_apps = []
unique_apps = []

for app in gps_data[1:]:
    name = app[0]
    if name in unique_apps:
        duplicate_apps.append(name)
    else:
        unique_apps.append(name)
        
print('Number of duplicates apps:', len(duplicate_apps))
print('\n')
print('Examples of duplicate apps:', duplicate_apps[:15])


Number of duplicates apps: 1181


Examples of duplicate apps: ['Quick PDF Scanner + OCR FREE', 'Box', 'Google My Business', 'ZOOM Cloud Meetings', 'join.me - Simple Meetings', 'Box', 'Zenefits', 'Google Ads', 'Google My Business', 'Slack', 'FreshBooks Classic', 'Insightly CRM', 'QuickBooks Accounting: Invoicing & Expenses', 'HipChat - Chat Built for Teams', 'Xero Accounting Software']


We dont want to delete random duplicates, the fourth column describes the number of reviews, we would essentially want to keep the data with the most number of reviews because it can be seen as the most recent data set. 

In order to do this we will:
- Creat a dictionary where each key is a unique app name, value is the highest number of reviews of that app
- Use the dictionary to create a new data set, which will have only one entry per app

In [9]:
print('Expected length:', len(gps_data) - 1181)

Expected length: 9660


In [14]:
reviews_max = {}

for app in gps_data[1:]:
    name = app[0]
    n_reviews = float(app[3])
    
    if name in reviews_max and reviews_max[name] < n_reviews:
        reviews_max[name] = n_reviews
        
    elif name not in reviews_max:
        reviews_max[name] = n_reviews
        
print(len(reviews_max))

android_clean = []
already_added = []

for app in gps_data[1:]:
    name = app[0]
    n_reviews = float(app[3])
    
    if (reviews_max[name] == n_reviews) and (name not in already_added):
        android_clean.append(app)
        already_added.append(name)
        

9659


In [16]:
explore_data(android_clean, 0, 3, True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Number of rows: 9659
Number of columns: 13


Android Clean is the true data set that has non-duplicate entries.


## Removing Non-English Apps


In [24]:
#Function to check if string has english characters or not.

def check(string):
    
    for character in string:
        check = ord(character)
        if check > 0 and check < 127:
            print('True')
            return True
        else:
            print("False")
            return False
        
test1 = check('Instagram')
test2 = check('爱奇艺PPS -《欢乐颂2》电视剧热播')
test3 = check('Docs To Go™ Free Office Suite')
test4 = check('Instachat 😜')
        
    

True
False
True
True


In [29]:
def check(string):
    count = 0
    for character in string:
        check = ord(character)
        if check > 127:
            count+=1
    
    if (count > 3):
        return False
    else:
        return True

print(check('爱奇艺PPS -《欢乐颂2》电视剧热播'))
print(check('Docs To Go™ Free Office Suite'))
print(check('Instachat 😜'))

False
True
True


In [34]:
gps_english = []
app_english = []

for app in android_clean:
    name = app[0]
    if check(name):
        gps_english.append(app)

for app in app_data[1:]:
    name = app[1]
    if check(name):
        app_english.append(app)
        
explore_data(gps_english,0,3,True)
explore_data(app_english,0,3,True)
        

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Number of rows: 9614
Number of columns: 13
['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'G

### Filtered Data so far

gps_english and app_english have been filtered so far for incorrect, duplicate and english only apps


In [47]:
## Finding number of free aps


gps_free = []
ios_free = []

for app in gps_english:
    price = (app[7])
    
    if (price == '0'):
        gps_free.append(app)
        

print(len(gps_free))


for app in app_english:
    price = (app[4])
    
    if (price == '0.0'):
        ios_free.append(app)
        
print(len(ios_free))

8864
3222


Finalized Data is :

- Google Playstore Data - 8864 Entries
- Apple Store Data - 3222 Entries


## Building Frequency Tables


In [55]:
def freq (dataset, index):
    table = {}
    total = 0
    
    for row in dataset:
        total += 1
        value = row[index]
        if value in table:
            table[value] += 1
        else:
            table[value] = 1
            
    table_percentages = {}
    for key in table:
        percentage = (table[key] / total) * 100
        table_percentages[key] = percentage
    
    return table_percentages

def display (dataset,index):
    table = freq(dataset, index)
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key)
        table_display.append(key_val_as_tuple)
        
    table_sorted = sorted(table_display, reverse = True)
    for entry in table_sorted:
        print(entry[1], ':', entry[0])
        
        
display(ios_free, -5)
        

Games : 58.16263190564867
Entertainment : 7.883302296710118
Photo & Video : 4.9658597144630665
Education : 3.662321539416512
Social Networking : 3.2898820608317814
Shopping : 2.60707635009311
Utilities : 2.5139664804469275
Sports : 2.1415270018621975
Music : 2.0484171322160147
Health & Fitness : 2.0173805090006205
Productivity : 1.7380509000620732
Lifestyle : 1.5828677839851024
News : 1.3345747982619491
Travel : 1.2414649286157666
Finance : 1.1173184357541899
Weather : 0.8690254500310366
Food & Drink : 0.8069522036002483
Reference : 0.5586592178770949
Business : 0.5276225946617008
Book : 0.4345127250155183
Navigation : 0.186219739292365
Medical : 0.186219739292365
Catalogs : 0.12414649286157665


### Genre - App Store

The most common primary genre for the app store is games with a 58.16 % majority over the rest of the genres


In [56]:
display(gps_free, 1)

FAMILY : 18.907942238267147
GAME : 9.724729241877256
TOOLS : 8.461191335740072
BUSINESS : 4.591606498194946
LIFESTYLE : 3.9034296028880866
PRODUCTIVITY : 3.892148014440433
FINANCE : 3.7003610108303246
MEDICAL : 3.531137184115524
SPORTS : 3.395758122743682
PERSONALIZATION : 3.3167870036101084
COMMUNICATION : 3.2378158844765346
HEALTH_AND_FITNESS : 3.0798736462093865
PHOTOGRAPHY : 2.944494584837545
NEWS_AND_MAGAZINES : 2.7978339350180503
SOCIAL : 2.6624548736462095
TRAVEL_AND_LOCAL : 2.33528880866426
SHOPPING : 2.2450361010830324
BOOKS_AND_REFERENCE : 2.1435018050541514
DATING : 1.861462093862816
VIDEO_PLAYERS : 1.7937725631768955
MAPS_AND_NAVIGATION : 1.3989169675090252
FOOD_AND_DRINK : 1.2409747292418771
EDUCATION : 1.1620036101083033
ENTERTAINMENT : 0.9589350180505415
LIBRARIES_AND_DEMO : 0.9363718411552346
AUTO_AND_VEHICLES : 0.9250902527075812
HOUSE_AND_HOME : 0.8235559566787004
WEATHER : 0.8009927797833934
EVENTS : 0.7107400722021661
PARENTING : 0.6543321299638989
ART_AND_DESIGN : 

### Category - Google Play Store

The most common category in the google play store is the family genre with a 18.90 percentage

In [60]:
display(gps_free, -4)

Tools : 8.449909747292418
Entertainment : 6.069494584837545
Education : 5.347472924187725
Business : 4.591606498194946
Productivity : 3.892148014440433
Lifestyle : 3.892148014440433
Finance : 3.7003610108303246
Medical : 3.531137184115524
Sports : 3.463447653429603
Personalization : 3.3167870036101084
Communication : 3.2378158844765346
Action : 3.1024368231046933
Health & Fitness : 3.0798736462093865
Photography : 2.944494584837545
News & Magazines : 2.7978339350180503
Social : 2.6624548736462095
Travel & Local : 2.3240072202166067
Shopping : 2.2450361010830324
Books & Reference : 2.1435018050541514
Simulation : 2.0419675090252705
Dating : 1.861462093862816
Arcade : 1.8501805054151623
Video Players & Editors : 1.7712093862815883
Casual : 1.7599277978339352
Maps & Navigation : 1.3989169675090252
Food & Drink : 1.2409747292418771
Puzzle : 1.128158844765343
Racing : 0.9927797833935018
Role Playing : 0.9363718411552346
Libraries & Demo : 0.9363718411552346
Auto & Vehicles : 0.9250902527075

### Category - Google Play Store

Tools is the highest with 8.44

## Most Popular Apps by Genre on the App Store


In [62]:
genres_ios = freq(ios_free, -5)

for genre in genres_ios:
    total = 0
    len_genre = 0
    for app in ios_free:
        genre_app = app[-5]
        if genre_app == genre:
            n_ratings = float(app[5])
            total += n_ratings
            len_genre += 1
    avg_n_ratings = total / len_genre
    print(genre, ':', avg_n_ratings)

Photo & Video : 28441.54375
Education : 7003.983050847458
Entertainment : 14029.830708661417
Lifestyle : 16485.764705882353
News : 21248.023255813954
Finance : 31467.944444444445
Navigation : 86090.33333333333
Travel : 28243.8
Health & Fitness : 23298.015384615384
Business : 7491.117647058823
Shopping : 26919.690476190477
Social Networking : 71548.34905660378
Weather : 52279.892857142855
Catalogs : 4004.0
Games : 22788.6696905016
Music : 57326.530303030304
Food & Drink : 33333.92307692308
Book : 39758.5
Productivity : 21028.410714285714
Sports : 23008.898550724636
Utilities : 18684.456790123455
Reference : 74942.11111111111
Medical : 612.0


In [64]:
for app in ios_free:
    if app[-5] == 'Navigation':
        print(app[1], ':', app[5]) #Print name and number of ratings

Waze - GPS Navigation, Maps & Real-time Traffic : 345046
Google Maps - Navigation & Transit : 154911
Geocaching® : 12811
CoPilot GPS – Car Navigation & Offline Maps : 3582
ImmobilienScout24: Real Estate Search in Germany : 187
Railway Route Search : 5


In [69]:
for app in ios_free:
    if app[-5] == 'Reference':
        print(app[1], ':', app[5])

Bible : 985920
Dictionary.com Dictionary & Thesaurus : 200047
Dictionary.com Dictionary & Thesaurus for iPad : 54175
Google Translate : 26786
Muslim Pro: Ramadan 2017 Prayer Times, Azan, Quran : 18418
New Furniture Mods - Pocket Wiki & Game Tools for Minecraft PC Edition : 17588
Merriam-Webster Dictionary : 16849
Night Sky : 12122
City Maps for Minecraft PE - The Best Maps for Minecraft Pocket Edition (MCPE) : 8535
LUCKY BLOCK MOD ™ for Minecraft PC Edition - The Best Pocket Wiki & Mods Installer Tools : 4693
GUNS MODS for Minecraft PC Edition - Mods Tools : 1497
Guides for Pokémon GO - Pokemon GO News and Cheats : 826
WWDC : 762
Horror Maps for Minecraft PE - Download The Scariest Maps for Minecraft Pocket Edition (MCPE) Free : 718
VPN Express : 14
Real Bike Traffic Rider Virtual Reality Glasses : 8
教えて!goo : 0
Jishokun-Japanese English Dictionary & Translator : 0


## Most Popular Apps by Genre on Google Play

In [72]:
display(gps_free,5) #The Installs Columns

1,000,000+ : 15.726534296028879
100,000+ : 11.552346570397113
10,000,000+ : 10.548285198555957
10,000+ : 10.198555956678701
1,000+ : 8.393501805054152
100+ : 6.915613718411552
5,000,000+ : 6.825361010830325
500,000+ : 5.561823104693141
50,000+ : 4.7721119133574
5,000+ : 4.512635379061372
10+ : 3.5424187725631766
500+ : 3.2490974729241873
50,000,000+ : 2.3014440433213
100,000,000+ : 2.1322202166064983
50+ : 1.917870036101083
5+ : 0.78971119133574
1+ : 0.5076714801444043
500,000,000+ : 0.2707581227436823
1,000,000,000+ : 0.22563176895306858
0+ : 0.04512635379061372
0 : 0.01128158844765343


In [74]:
categories_android = freq(gps_free,1)

for category in categories_android:
    total = 0
    len_category = 0
    for app in gps_free:
        category_app = app[1]
        if category_app == category:
            n_installs = app[5]
            n_installs = n_installs.replace(',', '')
            n_installs = n_installs.replace('+', '')
            total += float(n_installs)
            len_category += 1
    avg_n_installs = total / len_category
    print(category, ':', avg_n_installs)
    

PARENTING : 542603.6206896552
SOCIAL : 23253652.127118643
TOOLS : 10801391.298666667
LIBRARIES_AND_DEMO : 638503.734939759
BUSINESS : 1712290.1474201474
HOUSE_AND_HOME : 1331540.5616438356
FINANCE : 1387692.475609756
TRAVEL_AND_LOCAL : 13984077.710144928
AUTO_AND_VEHICLES : 647317.8170731707
DATING : 854028.8303030303
LIFESTYLE : 1437816.2687861272
MAPS_AND_NAVIGATION : 4056941.7741935486
WEATHER : 5074486.197183099
FAMILY : 3695641.8198090694
ENTERTAINMENT : 11640705.88235294
COMMUNICATION : 38456119.167247385
MEDICAL : 120550.61980830671
SPORTS : 3638640.1428571427
HEALTH_AND_FITNESS : 4188821.9853479853
PRODUCTIVITY : 16787331.344927534
EDUCATION : 1833495.145631068
FOOD_AND_DRINK : 1924897.7363636363
SHOPPING : 7036877.311557789
PERSONALIZATION : 5201482.6122448975
PHOTOGRAPHY : 17840110.40229885
COMICS : 817657.2727272727
GAME : 15588015.603248259
BEAUTY : 513151.88679245283
BOOKS_AND_REFERENCE : 8767811.894736841
NEWS_AND_MAGAZINES : 9549178.467741935
VIDEO_PLAYERS : 24727872.452